closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -1070,23 +887,615 @@ boolean isRestrictedProviderAllowed(String providerName) {
+                    String[] attributeArray = inAttributes.split(":");
+                    for (String attribute : attributeArray) {
+                        String[] in = attribute.split("=", 2);
+                        if (in.length != 2) {
+                            printStackTraceAndExit(
+                                    "Constraint attributes format is incorrect: " + providerInfo);
+                        }
+                    }
+                }
+                Constraint constraint = new Constraint(inType, inAlgorithm, inAttributes);
+                constraints.add(constraint);
+            }
+
+            // Differeriante between add, remove and override.
+            if (!isNullOrBlank(action)) {
+                if (providerChanged) {
+                    printStackTraceAndExit(
+                        "Cannot append or remove constraints since the provider " + providerName
+                        + " wasn't in this position in the profile extended.");
+                }
+                List<Constraint> existingConstraints = providerConstraints.get(providerName);
+                if (existingConstraints == null) {
+                    existingConstraints = new ArrayList<>();
+                    providerConstraints.put(providerName, existingConstraints);
+                }
+                if (action.equals("+")) { // Appending constraints.
+                    existingConstraints.addAll(constraints);
+                } else { // Removing constraints.
+                    for (Constraint toRemove: constraints) {
+                        if (!existingConstraints.remove(toRemove)) {
+                            printStackTraceAndExit(
+                                    "Constraint " + toRemove + "is not part of existing constraints.");
+                        }
+                    }
+                }
+            } else {
+                providerConstraints.put(providerName, constraints);
+            }
+
+            if (debug != null) {
+                debug.println("\t\t\tSuccessfully loaded constraints for security provider: " + providerName);
+            }
+        }
+
+        private void checkHexValues() {
+            for (String profileID: profilesHashes.keySet()) {
+                String hashValue = profilesHashes.get(profileID);
+                List<String> allInfo = profilesInfo.get(profileID);
+
+                if (debug != null) {
+                    debug.println("Calculating hash for '" + profileID + "'.");
+                }
+                String[] hashInfo = hashValue.split(":");
+                String digestAlgo = hashInfo[0].trim();
+                String expectedHash = hashInfo[1].trim();
+                try {
+                    MessageDigest md = MessageDigest.getInstance(digestAlgo);
+                    byte[] allInfoArray = allInfo.stream().sorted().collect(Collectors.joining("\n")).getBytes();
+                    byte[] resultHashArray = md.digest(allInfoArray);
```

Comment:
This should use a specific charset, say, `StandardCharsets.UTF_8`.
How does this avoid including the `.hash` property in the computation?

Reply:
Not really sure what you mean about the `StandardCharsets.UTF_8`.

The `.hash` property is not added to the `StringBuilder`, so it's not included in the computation.

Reply:
We need a consistent and predictable encoding of the string (using `getByte()` involves `Charset.defaultCharset()` which is affected by the `file.encoding` system property):
```java
                    byte[] allInfoArray = allInfo.stream().sorted()
                            .collect(Collectors.joining("\n"))
                            .getBytes(StandardCharsets.UTF_8);
```

Reply:
Changed.